src/Chapter3.hs

```diff
@@ -1006,6 +1129,118 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data KAction = KAttack | KPotion | KSpell
+
+data K = MkK
+  { kHealth :: Health
+  , kAttack :: Attack
+  , kDefense :: Defense
+  , kActions :: [KAction]
+  , usedSpell :: Bool
+  , usedPotion :: Bool
+  }
+
+mkK :: Health -> Attack -> Defense -> [KAction] -> K
+mkK h a d k =
+  MkK { kHealth = h
+      , kAttack = a
+      , kDefense = d
+      , kActions = cycle k
+      , usedSpell = False
+      , usedPotion = False
+      }
+
+data MAction = MAttack | MFlee
+data M = MkM
+  { mHealth :: Health
+  , mAttack :: Attack
+  , mActions :: [MAction]
+  }
+
+mkM :: Health -> Attack -> [MAction] -> M
+mkM h a m =
+  MkM { mHealth = h
+      , mAttack = a
+      , mActions = cycle m
+      }
+
+data Turn = TAttack Damage | TIdle | TFlee
+
+class Fighter a where
+  damage :: a -> Damage
+  getDamaged :: a -> Damage -> a
+  isDead :: a -> Bool
+  nextAction :: a -> (Turn, a)
+
+instance Fighter K where
+  damage :: K -> Damage
+  damage k = Damage (a + (if usedSpell k then 1 else 0))
+    where Attack a = kAttack k
+
+  getDamaged :: K -> Damage -> K
+  getDamaged k (Damage d) =
+    let Health kHealth' = kHealth k in
+    let Defense kDefense' = kDefense k in
+    k { kHealth = Health (min (kHealth' + kDefense' + (if usedPotion k then 1 else 0) - d) kHealth') }
+
+  isDead :: K -> Bool
+  isDead k = h <= 0
+    where Health h = kHealth k
+
+  nextAction :: K -> (Turn, K)
+  nextAction k =
+    let action = head $ kActions k
+        remaining = drop 1 $ kActions k
+        (action', k') = case action of
+          KAttack -> (TAttack (damage k), k)
+          KPotion -> (TIdle, k { usedPotion = True })
+          KSpell -> (TIdle, k { usedSpell = True })
+    in
+    (action', k' { kActions = remaining })
+
+instance Fighter M where
+  damage :: M -> Damage
+  damage m = Damage a
+    where Attack a = mAttack m
+
+  getDamaged m (Damage d) =
+    let Health mHealth' = mHealth m in
+    m { mHealth = Health (mHealth' - d) }
+
+  isDead :: M -> Bool
+  isDead m = h <= 0
+    where Health h = mHealth m
+
+  nextAction :: M -> (Turn, M)
+  nextAction m =
+    let action = head $ mActions m
+        remaining = drop 1 $ mActions m
+        action' = case action of
+          MAttack -> TAttack (damage m)
+          MFlee -> TFlee
+    in
+    (action', m { mActions = remaining })
+
+
+winner :: (Fighter a) => a -> a -> a
```

Comment:
Unfortunately, this type signature doesn't allow fighting of warriors of different types :disappointed: 

Here you want something like this:

```haskell
winner :: (Fighter a, Fighter b) => a -> b -> ...
```